langkit/diagnostics.py

```diff
@@ -65,77 +68,46 @@ def set_style(cls, style):
         cls.style = style
 
 
+@dataclass(order=True, frozen=True)
 class Location:
     """
     Holder for a location in the source code.
     """
 
-    def __init__(self, file, line, column=0):
-        self.file = file
-        """
-        Path to the file for this location.
-
-        :type: str
-        """
-
-        self.line = line
-        """
-        Location line number (1-based).
-
-        :type: int
-        """
-
-        self.column = column
-        """
-        Column number (1-based). Zero if unspecified.
-
-        :type: int
-        """
-
-        self.previous_in_callstack = None
-        """
-        If this location is created in the context of a call stack, link to the
-        location of the previous (i.e. caller) frame in this stack. Use the
-        `short_repr` property to format the whole call stack.
-
-        :type: None|Location
-        """
+    file: str
+    """
+    Path to the file for this location.
+    """
 
-    @property
-    def as_tuple(self):
-        return (self.file, self.line, self.column)
+    line: int
+    """
+    Line number (1-based).
+    """
 
-    def __eq__(self, other):
-        return self.as_tuple == other.as_tuple
+    column: int = field(default=0)
+    """
+    Column number (1-based). Zero if unspecified.
 
-    def __lt__(self, other):
-        return self.as_tuple < other.as_tuple
+    :type: int
+    """
 
-    def __hash__(self):
-        return hash(self.as_tuple)
+    end_line: int = field(default=0)
+    end_column: int = field(default=0)
+    """
+    End line and column numbers. Zero if unspecified.
+    """
 
-    def __repr__(self):
-        return '<Location {} {}:{}>'.format(self.file, self.line, self.column)
+    # RA22-015 TODO: Remove this "zero if unspecified" business when we get rid
+    # of the legacy DSL.
 
     @property
-    def short_repr(self):
+    def gnu_style_repr(self) -> str:
         """
-        Return a representation for this location (and its callers, if this
-        info is present) as a human readable string.
+        Return a GNU style representation for this Location, in the form::
 
-        :rtype: str
+            file:line:column
         """
-        # Reconstruct the call stack
-        stack = []
-        loc = self
-        while loc is not None:
-            loc_str = '{}:{}'.format(P.basename(loc.file), loc.line)
-            if loc.column:
-                loc_str += ':{}'.format(loc.column)
-            stack.append(loc_str)
-            loc = loc.previous_in_callstack
-
-        return '[{}]'.format(', '.join(stack))
+        return "{}:{}:{}".format(self.file, self.line, self.column)
```

Comment:
GNU tools don’t print the column number when it is unknown, and that’s what the previous implementation did. Was this change intentional?

Reply:
No it's a mistake, good catch!

Reply:
It’s not fixed, yet: we can see `:0: ` patterns in the rewritten baselines